iroh-gossip/src/net/handles.rs

```diff
@@ -0,0 +1,254 @@
+            .send(Command::BroadcastNeighbors(message))
+            .await
+            .map_err(|_| anyhow!("Gossip actor dropped"))
+    }
+
+    /// Join a set of peers.
+    pub async fn join_peers(&self, peers: Vec<NodeId>) -> anyhow::Result<()> {
+        self.0
+            .send(Command::JoinPeers(peers))
+            .await
+            .map_err(|_| anyhow!("Gossip actor dropped"))
+    }
+}
+
+type EventStream = Pin<Box<dyn Stream<Item = Result<Event>> + Send + 'static>>;
+
+/// Subscribed gossip topic.
+///
+/// This handle is a [`Stream`] of [`Event`]s from the topic, and can be used to send messages.
+///
+/// It may be split into sender and receiver parts with [`Self::split`].
+#[derive(Debug)]
+pub struct GossipTopic {
+    sender: GossipSender,
+    receiver: GossipReceiver,
+}
+
+impl GossipTopic {
+    pub(crate) fn new(sender: async_channel::Sender<Command>, receiver: EventStream) -> Self {
+        Self {
+            sender: GossipSender::new(sender),
+            receiver: GossipReceiver::new(Box::pin(receiver)),
+        }
+    }
+
+    /// Splits `self` into [`GossipSender`] and [`GossipReceiver`] parts.
+    pub fn split(self) -> (GossipSender, GossipReceiver) {
+        (self.sender, self.receiver)
+    }
+
+    /// Sends a message to all peers.
+    pub async fn broadcast(&self, message: Bytes) -> anyhow::Result<()> {
+        self.sender.broadcast(message).await
+    }
+
+    /// Sends a message to our direct neighbors in the swarm.
+    pub async fn broadcast_neighbors(&self, message: Bytes) -> anyhow::Result<()> {
+        self.sender.broadcast_neighbors(message).await
+    }
+
+    /// Waits until we are connected to at least one node.
+    pub async fn joined(&mut self) -> Result<()> {
+        self.receiver.joined().await
+    }
+
+    /// Returns true if we are connected to at least one node.
+    pub fn is_joined(&self) -> bool {
+        self.receiver.is_joined()
+    }
+}
+
+impl Stream for GossipTopic {
+    type Item = Result<Event>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.receiver).poll_next(cx)
+    }
+}
+
+/// Receiver for gossip events on a topic.
+///
+/// This is a [`Stream`] of [`Event`]s emitted from the topic.
+#[derive(derive_more::Debug)]
+pub struct GossipReceiver {
+    #[debug("EventStream")]
+    stream: EventStream,
+    neighbors: HashSet<NodeId>,
+}
+
+impl GossipReceiver {
+    pub(crate) fn new(events_rx: EventStream) -> Self {
+        Self {
+            stream: events_rx,
+            neighbors: Default::default(),
+        }
+    }
+
+    /// Lists our current direct neighbors.
+    pub fn neighbors(&self) -> impl Iterator<Item = NodeId> + '_ {
+        self.neighbors.iter().copied()
+    }
+
+    /// Waits until we are connected to at least one node.
+    pub async fn joined(&mut self) -> Result<()> {
+        while self.neighbors.is_empty() {
+            let _ = self.try_next().await?;
```

Comment:
shouldn't this be explicit about the type of event?

Reply:
the `Joined` event is handled in `try_next`. But we could be more explict and return an error if any other event than `Joined` is returned, yes.